src/Data/Variant.purs

```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
+      M.Nothing
+        | tagIx == 0 → M.Nothing
+        | otherwise →
+          map (\p → VariantRep { type: T.fst p, value: (T.fst $ T.snd p).top })
+          $ L.index zipped $ tagIx - 1
+        where
+        prx = RLProxy ∷ RLProxy rl
+        tags = variantTags prx
+        dicts = variantBoundedEnums prx
+        bounds = variantBounded prx
+        zipped = L.zip tags $ L.zip bounds dicts
+
+        coerceOut ∷ M.Maybe (VariantRep VariantCase) → M.Maybe (Variant r)
+        coerceOut = unsafeCoerce
+
+        VariantRep arep = unsafeCoerce a
+
+        tagIx ∷ Int
+        tagIx = fromJust "enum" $ L.findIndex (\(T.Tuple s _) → arep.type == s) zipped
+
+        pack = fromJust "enum" $ L.index zipped tagIx
+
+        dict ∷ BoundedEnumDict VariantCase
+        dict = T.snd $ T.snd pack
+
+    succ a = coerceOut case dict.succ (unsafeCoerce arep.value) of
+      M.Just b → M.Just $ VariantRep arep{ value = b }
+      M.Nothing
+        | tagIx == totalTags - 1 → M.Nothing
+        | otherwise →
+          map (\p → VariantRep { type: T.fst p, value: (T.fst $ T.snd p).bottom })
+          $ L.index zipped $ tagIx + 1
+        where
+        prx = RLProxy ∷ RLProxy rl
+        tags = variantTags prx
+        dicts = variantBoundedEnums prx
+        bounds = variantBounded prx
+        zipped = L.zip tags $ L.zip bounds dicts
+
+        coerceOut ∷ M.Maybe (VariantRep VariantCase) → M.Maybe (Variant r)
+        coerceOut = unsafeCoerce
+
+        VariantRep arep = unsafeCoerce a
+
+        tagIx ∷ Int
+        tagIx = fromJust "enum" $ L.findIndex (\(T.Tuple s _) → arep.type == s) zipped
+
+        pack = fromJust "enum" $ L.index zipped tagIx
+
+        dict ∷ BoundedEnumDict VariantCase
+        dict = T.snd $ T.snd pack
+
+        totalTags = L.length tags
+
+
+instance boundedEnumVariant
+  ∷ ( R.RowToList r rl
+    , VariantTags rl
+    , VariantEqs rl
+    , VariantOrds rl
+    , VariantBoundedEnums rl
+    )
+  ⇒ BoundedEnum (Variant r) where
+    cardinality =
+      Cardinality
+      $ F.sum
+      $ map _.cardinality
+      $ variantBoundedEnums (RLProxy ∷ RLProxy rl)
+
+    fromEnum inp = go 0 packs
+      where
+      packs =
+        L.sortBy (\(T.Tuple a _) (T.Tuple b _) → compare a b)
+        $ L.zip
+          (variantTags (RLProxy ∷ RLProxy rl))
+          (variantBoundedEnums (RLProxy ∷ RLProxy rl))
+
+      VariantRep arep = unsafeCoerce inp
+
+      tagIx =
+        fromJust "boundedEnum"
+        $ L.findIndex (\(T.Tuple s _) → s == arep.type) packs
+
+      pack = fromJust "enum" $ L.index packs tagIx
+
+      go ∷ Int → L.List (T.Tuple String (BoundedEnumDict VariantCase)) → Int
+      go acc = case _ of
+        L.Cons (T.Tuple k dict) tl
+          | k == arep.type →
+              acc + dict.fromEnum arep.value
+          | otherwise →
+              let c = dict.cardinality
+              in go (acc + c) tl
+        _ → unsafeCrashWith "Data.Variant: impossible boundedEnum"
```

Comment:
Similar issues. Unnecessary sorting, we don't need the additional zip, `tagIx` and `pack` are unused.